proxy_list.go

```diff
@@ -0,0 +1,332 @@
+
+List is not a partitioned Hazelcast data structure. So all the contents of the List are stored in a single
+machine (and in the backup). So, a single List will not scale by adding more members in the cluster.
+*/
+type List struct {
+	*proxy
+	partitionID int32
+}
+
+func newList(p *proxy) (*List, error) {
+	if partitionID, err := p.stringToPartitionID(p.name); err != nil {
+		return nil, err
+	} else {
+		return &List{proxy: p, partitionID: partitionID}, nil
+	}
+}
+
+// AddListener adds an item listener for this list.
+// The listener will be invoked whenever an item is added to or removed from this list.
+// Returns subscription ID of the listener.
+func (l *List) AddListener(handler ListItemNotifiedHandler) (types.UUID, error) {
+	return l.addListener(false, handler)
+}
+
+// AddListener adds an item listener for this list.
+// The listener will be invoked whenever an item is added to or removed from this list.
+// Received events include the updated item.
+// Returns subscription ID of the listener.
+func (l *List) AddListenerIncludeValue(handler ListItemNotifiedHandler) (types.UUID, error) {
+	return l.addListener(true, handler)
+}
+
+func (l *List) addListener(includeValue bool, handler ListItemNotifiedHandler) (types.UUID, error) {
+	subscriptionID := types.NewUUID()
+	addRequest := codec.EncodeListAddListenerRequest(l.name, includeValue, l.config.ClusterConfig.SmartRouting)
+	removeRequest := codec.EncodeListRemoveListenerRequest(l.name, subscriptionID)
+	listenerHandler := func(msg *proto.ClientMessage) {
+		codec.HandleListAddListener(msg, func(itemData serialization.Data, uuid types.UUID, eventType int32) {
+			item, err := l.convertToObject(itemData)
+			if err != nil {
+				l.logger.Warnf("cannot convert data to Go value: %v", err)
+				return
+			}
+			member := l.clusterService.GetMemberByUUID(uuid.String())
+			handler(newListItemNotified(l.name, item, member, eventType))
+		})
+	}
+	err := l.listenerBinder.Add(subscriptionID, addRequest, removeRequest, listenerHandler)
+	return subscriptionID, err
+}
+
+// RemoveListener removes the item listener with the given subscription ID.
+func (l *List) RemoveListener(subscriptionID types.UUID) error {
+	return l.listenerBinder.Remove(subscriptionID)
+}
+
+// Add appends the specified element to the end of this list.
+// Returns true if the list has changed as a result of this operation, false otherwise.
+func (l *List) Add(element interface{}) (bool, error) {
+	elementData, err := l.validateAndSerialize(element)
+	if err != nil {
+		return false, err
+	}
+	request := codec.EncodeListAddRequest(l.name, elementData)
+	response, err := l.invokeOnPartition(context.TODO(), request, l.partitionID)
+	if err != nil {
+		return false, err
+	}
+	return codec.DecodeListAddResponse(response), nil
+}
+
+// AddAt inserts the specified element at the specified index.
+// Shifts the subsequent elements to the right.
+func (l *List) AddAt(index int32, element interface{}) error {
```

Comment:
I think `index` should be `int` instead of `int32`. Although the API takes an `int32`, an `int` is more appropriate to denote an index, size, etc.

Reply:
Java's `j.u.List` supports signed 32-bit integer indexes, so in case if we accept `int` users may face overflow issues. I think, it's better to be precise about the expected number format to avoid such issues.